lib/spack/spack/cmd/install.py

```diff
@@ -259,12 +263,13 @@ def install(parser, args, **kwargs):
         env = ev.get_env(args, 'install')
         if env:
             if not args.only_concrete:
-                concretized_specs = env.concretize()
-                ev.display_specs(concretized_specs)
+                with env.write_transaction():
+                    concretized_specs = env.concretize()
+                    ev.display_specs(concretized_specs)
 
-                # save view regeneration for later, so that we only do it
-                # once, as it can be slow.
-                env.write(regenerate_views=False)
+                    # save view regeneration for later, so that we only do it
+                    # once, as it can be slow.
+                    env.write(regenerate_views=False)
```

Comment:
This one's inside the write transaction while the one above is not -- which should it be?

Reply:
As far as I can see, all calls to `Environment.write` that occur inside `cmd.install` are inside a `Environment.write_transaction` and all calls to `Environment.regenerate_views` are not. Are you saying you see something different? Or are you saying there is some other sort of inconsistency?

Reply:
nevermind - I must've misread the indentation in the diff.  When I check this out, they are indeed both outside.

I guess my remaining question would be -- are we sure that concurrently calling regenerate_views is safe?  Does every call regenerate links for all specs?

Reply:
RE: 
> I should have also clarified: I was operating on the assumption that it should be OK for multiple processes to run regenerate_views at the same time (though that may be expensive).

This is what I'm curious about.  Is it not redundant to regenerate links for all packages every time?  Seems like this operation should be locked or separated (I know I said it should be lock-free -- I was imagining that each spec would be linked in independently after it was installed).

Reply:
It is redundant but I don't think it can create errors. Nevertheless, to avoid redundant operations I decided to wrap the call to `Environment.regenerate_views` in an `Environment.write_transaction`.

Reply:
@scheibelp: if we're wrapping it in a write transaction, it's still redundant -- it's just redundant and serialized now.  Is there a way to avoid rewriting the view n times for n processes?

Reply:
I think the right way to do this is to hash the lockfile contents (which determine what is in the view), use a write transaction, and track the hash of the generated view w.r.t. the lockfile.  If the hash changes, we regenerate.  If not, we don't.  That should probably be a follow-on PR of its own.

Reply:
`ViewDescriptor.regenerate` only generates symlinks for packages that have not been added to the view (i.e. it doesn't blow everything away and restart from scratch), so I think wrapping each call to `regenerate_views` in a `write_transaction` will handle a large portion of the redundancy.

Reply:
@scheibelp: ah yes -- the view tracks what is installed.  Ok, in that case I think we are probably good and the write transaction will help.  We could do finer-grained locking on individual packages *in* the view later if we need to.  Thanks!

Reply:
@scheibelp: ah yes -- the view tracks what is installed. Ok, in that case I think we are probably good and the write transaction will help. We could do finer-grained locking on individual packages in the view later if we need to. Thanks!